tracer/src/Datadog.Trace/Iast/DefaultTaintedMap.cs

```diff
@@ -0,0 +1,244 @@
+            // We do not control duplicate entries.
+            _map.TryGetValue(index, out var existingValue);
+            entry.Next = existingValue;
+
+            // We only count the entries if we are not in flat mode
+            Interlocked.Increment(ref _entriesCount);
+        }
+
+        // If we flipped to flat mode:
+        // - Always override elements ignoring chaining.
+        // - Stop updating the estimated size.
+
+        _map[index] = entry;
+
+        if ((entry.PositiveHashCode & PurgeMask) == 0)
+        {
+            Purge();
+        }
+    }
+
+    /// <summary>
+    /// Purge entries that have been garbage collected. Only one concurrent call to this method is
+    /// allowed, further concurrent calls will be ignored.
+    /// </summary>
+    internal void Purge()
+    {
+        // Ensure we enter only once concurrently.
+        lock (_purgingLock)
+        {
+            if (_isPurging)
+            {
+                return;
+            }
+
+            _isPurging = true;
+        }
+
+        try
+        {
+            // Remove GC'd entries.
+            var removedCount = RemoveDeadKeys();
+
+            if (!IsFlat)
+            {
+                // We only count the entries if we are not in flat mode
+                if (Interlocked.Add(ref _entriesCount, -removedCount) > _flatModeThreshold)
+                {
+                    IsFlat = true;
+                }
+            }
+        }
+        finally
+        {
+            // Reset purging flag.
+            lock (_purgingLock)
+            {
+                _isPurging = false;
+            }
+        }
+    }
+
+    private int RemoveDeadKeys()
+    {
+        var removed = 0;
+        List<int> deadKeys = new();
+        ITaintedObject? previous;
+
+        foreach (var key in _map.Keys)
+        {
+            var current = _map[key];
+            previous = null;
+
+            while (current is not null)
+            {
+                if (!current.IsAlive)
+                {
+                    if (previous is null)
+                    {
+                        // We can delete the map key
+                        if (current.Next is null)
+                        {
+                            deadKeys.Add(key);
+                        }
+                        else
+                        {
+                            _map[key] = current.Next;
+                        }
+                    }
+                    else
+                    {
+                        previous.Next = current.Next;
+                    }
+
+                    current = current.Next;
+                    removed++;
+                }
+                else
+                {
+                    previous = current;
+                    current = current.Next;
+                }
+            }
+        }
+
+        foreach (var key in deadKeys)
+        {
+            _map.TryRemove(key, out _);
```

Comment:
If this returns false, should we change the value returned in `removed`? 🤔  I don't think so, just checking

Reply:
I agree. removed has a value indicating the number of taintedObjects removed. A dead key not being removed doesn't affect to the number of taintedObjects removed.